File: ZipReaderLibC/Zip.cpp
#include "Zip.h"
#include "ZipHeader.h"
#include "ZipExtraFieldRead.h"
#include "CodePage437.h"
#include <fstream>
using namespace std;


ZipReturn Zip::ZipFileOpen(const wchar_t* zipFilePath)
{
	_zipFs.open(zipFilePath, ios::binary | ios::in);

	ZipReturn ret = ZipFileReadHeaders();
	_zipFs.close();
	return ret;
}


ZipReturn Zip::ZipFileReadHeaders()
{
	ZipReturn zRet;
	zRet = FindEndOfCentralDirSignature();
	if (zRet != ZipGood)
	{
		return zRet;
	}

	long endOfCentralDir = (long)_zipFs.tellg();
	zRet = EndOfCentralDirRead();
	if (zRet != ZipGood)
	{
		return zRet;
	}

	// check if ZIP64 header is required
	bool zip64Required = (_centralDirStart == 0xffffffff || _centralDirSize == 0xffffffff || _localFilesCount == 0xffff);

	_zipFs.seekg(endOfCentralDir - 20, ios::beg);

	zRet = Zip64EndOfCentralDirectoryLocatorRead();
	if (zRet == ZipGood)
	{
		_zipFs.seekg((long)_endOfCentralDir64, ios::beg);
		zRet = Zip64EndOfCentralDirRead();
		if (zRet == ZipGood)
		{
			_zip64 = true;
			endOfCentralDir = _endOfCentralDir64;
		}
	}

	if (zip64Required && !_zip64)
		return Zip64EndOfCentralDirError;

	offset = (endOfCentralDir - _centralDirSize) - _centralDirStart;

	_centralDirStart += offset;

	// now read the central directory
	_zipFs.seekg(_centralDirStart, ios::beg);

	_centralDirectoryHeaders = new ZipHeader[_localFilesCount];
	for (unsigned int i = 0; i < _localFilesCount; i++)
	{
		zRet = CentralDirectoryHeaderReader(offset, &_centralDirectoryHeaders[i]);
		if (zRet != ZipGood)
			return zRet;
	}

	_localFileHeaders = new ZipHeader[_localFilesCount];
	for (unsigned int i = 0; i < _localFilesCount; i++)
	{
		zRet = LocalFileHeaderReader(_centralDirectoryHeaders[i].RelativeOffsetOfLocalHeader, _centralDirectoryHeaders[i].CompressedSize, &_localFileHeaders[i]);
		if (zRet != ZipGood)
			return zRet;
	}

	return ZipGood;
}





ZipReturn Zip::FindEndOfCentralDirSignature()
{
	_zipFs.seekg(0, ios::end);
	fileSize = (long)_zipFs.tellg();
	long maxBackSearch = 0xffff;

	if (fileSize < maxBackSearch)
	{
		maxBackSearch = fileSize;
	}

	const long buffSize = 0x400;

	unsigned char buffer[buffSize + 4];

	long backPosition = 4;
	while (backPosition < maxBackSearch)
	{
		backPosition += buffSize;
		if (backPosition > maxBackSearch)
		{
			backPosition = maxBackSearch;
		}

		long readSize = backPosition > buffSize + 4 ? buffSize + 4 : backPosition;

		_zipFs.seekg(fileSize - backPosition, ios::beg);
		_zipFs.read((char*)buffer, readSize);


		for (long i = readSize - 4; i >= 0; i--)
		{
			if (buffer[i] != 0x50 || buffer[i + 1] != 0x4b || buffer[i + 2] != 0x05 || buffer[i + 3] != 0x06)
			{
				continue;
			}

			_zipFs.seekg(fileSize - backPosition + i, ios::beg);

			return ZipGood;
		}
	}
	return ZipCentralDirError;
}


ZipReturn Zip::EndOfCentralDirRead()
{
	unsigned int thisSignature;
	_zipFs.read((char*)&thisSignature, 4);
	if (thisSignature != EndOfCentralDirSignature)
	{
		return ZipEndOfCentralDirectoryError;
	}

	unsigned short tUShort;
	_zipFs.read((char*)&tUShort, 2); // NumberOfThisDisk
	if (tUShort != 0)
	{
		return ZipEndOfCentralDirectoryError;
	}

	_zipFs.read((char*)&tUShort, 2); // NumberOfThisDiskCenterDir
	if (tUShort != 0)
	{
		return ZipEndOfCentralDirectoryError;
	}

	_zipFs.read((char*)&_localFilesCount, 2); // TotalNumberOfEntriesDisk

	_zipFs.read((char*)&tUShort, 2); // TotalNumber of entries in the central directory 
	if (tUShort != _localFilesCount)
	{
		return ZipEndOfCentralDirectoryError;
	}

	_zipFs.read((char*)&_centralDirSize, 4); // SizeOfCentralDir
	_zipFs.read((char*)&_centralDirStart, 4); // Offset

	unsigned short zipFileCommentLength;
	_zipFs.read((char*)&zipFileCommentLength, 2); // ZipFileCommentLength

	_fileComment = new unsigned char[zipFileCommentLength];
	_zipFs.read((char*)_fileComment, zipFileCommentLength); // ZipFileComment

	ExtraDataFoundOnEndOfFile = _zipFs.tellg() != fileSize;

	return ZipGood;
}


ZipReturn Zip::Zip64EndOfCentralDirRead()
{
	unsigned int thisSignature;
	_zipFs.read((char*)&thisSignature, 4);
	if (thisSignature != Zip64EndOfCentralDirSignature)
	{
		return ZipEndOfCentralDirectoryError;
	}

	unsigned long tULong;
	_zipFs.read((char*)&tULong, 8); // Size of zip64 end of central directory record
	if (tULong != 44)
	{
		return Zip64EndOfCentralDirError;
	}

	unsigned short tUShort;
	_zipFs.read((char*)&tUShort, 2); // version made by

	_zipFs.read((char*)&tUShort, 2); // version needed to extract
	if (tUShort != 45)
	{
		return Zip64EndOfCentralDirError;
	}

	unsigned int tUInt;
	_zipFs.read((char*)&tUInt, 4); // number of this disk
	if (tUInt != 0)
	{
		return Zip64EndOfCentralDirError;
	}

	_zipFs.read((char*)&tUInt, 4); // number of the disk with the start of the central directory
	if (tUInt != 0)
	{
		return Zip64EndOfCentralDirError;
	}

	_zipFs.read((char*)&_localFilesCount, 8); // total number of entries in the central directory on this disk

	_zipFs.read((char*)&tULong, 8); // total number of entries in the central directory
	if (tULong != _localFilesCount)
	{
		return Zip64EndOfCentralDirError;
	}

	_zip64 = true;
	_zipFs.read((char*)&_centralDirSize, 8); // size of the central directory

	_zipFs.read((char*)&_centralDirStart, 8); // offset of start of central directory with respect to the starting disk number

	return ZipGood;
}


ZipReturn Zip::Zip64EndOfCentralDirectoryLocatorRead()
{
	unsigned int thisSignature;
	_zipFs.read((char*)&thisSignature, 4);
	if (thisSignature != Zip64EndOfCentralDirectoryLocator)
	{
		return ZipEndOfCentralDirectoryError;
	}

	unsigned int tUInt;
	_zipFs.read((char*)&tUInt, 4); // number of the disk with the start of the zip64 end of central directory
	if (tUInt != 0)
	{
		return Zip64EndOfCentralDirectoryLocatorError;
	}

	_zipFs.read((char*)&_endOfCentralDir64, 8); // relative offset of the zip64 end of central directory record

	_zipFs.read((char*)&tUInt, 4); // total number of disks
	if (tUInt > 1)
	{
		return Zip64EndOfCentralDirectoryLocatorError;
	}

	return ZipGood;
}

ZipReturn Zip::CentralDirectoryHeaderReader(unsigned long long offset, ZipHeader* centralFile)
{
	unsigned int thisSignature;
	_zipFs.read((char*)&thisSignature, 4);
	if (thisSignature != CentralDirectoryHeaderSignature)
		return ZipCentralDirError;

	_zipFs.read((char*)&centralFile->VersionMadeBy, 2); // Version Made By
	_zipFs.read((char*)&centralFile->VersionNeededToExtract, 2); // Version Needed To Extract

	_zipFs.read((char*)&centralFile->GeneralPurposeBitFlag, 2); // General Purpose Bit Flag
	_zipFs.read((char*)&centralFile->CompressionMethod, 2); // Compression Method

	unsigned short lastModFileTime;
	_zipFs.read((char*)&lastModFileTime, 2); // Last Mod File Time

	unsigned short lastModFileDate;
	_zipFs.read((char*)&lastModFileDate, 2); // Last Mod File Date

	centralFile->HeaderLastModified = (lastModFileDate << 16) | lastModFileTime;

	_zipFs.read((char*)&centralFile->CRC, 4); // CRC

	unsigned int tUInt;
	_zipFs.read((char*)&tUInt, 4); // Compressed Size
	centralFile->CompressedSize = tUInt;

	_zipFs.read((char*)&tUInt, 4); // Uncompressed Size
	centralFile->UncompressedSize = tUInt;

	_zipFs.read((char*)&centralFile->fileNameLength, 2); // File Name Length
	_zipFs.read((char*)&centralFile->extraFieldLength, 2); // Extra Field Length
	_zipFs.read((char*)&centralFile->fileCommentLength, 2); // File Comment Length

	unsigned short tUShort;
	_zipFs.read((char*)&tUShort, 2); // Disk Number Start
	_zipFs.read((char*)&tUShort, 2); // InternalFileAttributes
	_zipFs.read((char*)&tUInt, 4); // ExternalFileAttributes

	_zipFs.read((char*)&centralFile->RelativeOffsetOfLocalHeader, 4); // Relative Offset Of Local Header

	centralFile->bFileNameHeader=new char[centralFile->fileNameLength];
	_zipFs.read((char*)centralFile->bFileNameHeader, centralFile->fileNameLength); // File Name

	//if bit 11 of the general purpose bit flag is set, the filename is encoded using UTF-8
	//if bit 11 of the general purpose bit flag is not set, the filename is encoded using Encoding.GetEncoding(437);
	if ((centralFile->GeneralPurposeBitFlag & (1 << 11)))
	{
		centralFile->bFileName = new char[centralFile->fileNameLength + 1];
		memcpy(centralFile->bFileName,centralFile->bFileNameHeader,centralFile->fileNameLength);
		centralFile->bFileName[centralFile->fileNameLength] = 0;
	}
	else
	{
		CodePage437 cp437=CodePage437();
		centralFile->bFileName = cp437.Enc(centralFile->bFileNameHeader, centralFile->fileNameLength);
	}

	centralFile->bExtraField = new char[centralFile->extraFieldLength];
	_zipFs.read((char*)centralFile->bExtraField, centralFile->extraFieldLength); // Extra Field

	if (centralFile->extraFieldLength > 0)
		ZipExtraFieldRead::ReadExtraField(centralFile, true);

	centralFile->RelativeOffsetOfLocalHeader += offset;

	centralFile->bFileComment = new char[centralFile->fileCommentLength];
	_zipFs.read((char*)centralFile->bFileComment, centralFile->fileCommentLength); // File Comment

	return ZipGood;
}

ZipReturn Zip::LocalFileHeaderReader(unsigned long long relativeOffsetOfLocalHeader, unsigned long long compressedSize, ZipHeader* localFile)
{
	localFile->RelativeOffsetOfLocalHeader = relativeOffsetOfLocalHeader;
	_zipFs.seekg((long)relativeOffsetOfLocalHeader, ios::beg);
	unsigned int thisSignature;
	_zipFs.read((char*)&thisSignature, 4);
	if (thisSignature != LocalFileHeaderSignature)
		return ZipCentralDirError;

	_zipFs.read((char*)&localFile->VersionNeededToExtract, 2); // Version Needed To Extract

	_zipFs.read((char*)&localFile->GeneralPurposeBitFlag, 2); // General Purpose Bit Flag
	_zipFs.read((char*)&localFile->CompressionMethod, 2); // Compression Method

	unsigned short lastModFileTime;
	_zipFs.read((char*)&lastModFileTime, 2); // Last Mod File Time

	unsigned short lastModFileDate;
	_zipFs.read((char*)&lastModFileDate, 2); // Last Mod File Date

	localFile->HeaderLastModified = (lastModFileDate << 16) | lastModFileTime;

	_zipFs.read((char*)&localFile->CRC, 4); // CRC

	unsigned int tUInt;
	_zipFs.read((char*)&tUInt, 4); // Compressed Size
	localFile->CompressedSize = tUInt;

	_zipFs.read((char*)&tUInt, 4); // Uncompressed Size
	localFile->UncompressedSize = tUInt;

	_zipFs.read((char*)&localFile->fileNameLength, 2); // File Name Length
	_zipFs.read((char*)&localFile->extraFieldLength, 2); // Extra Field Length
	localFile->fileCommentLength = 0;

	localFile->bFileNameHeader = new char[localFile->fileNameLength];
	_zipFs.read((char*)localFile->bFileNameHeader, localFile->fileNameLength); // File Name

	//if bit 11 of the general purpose bit flag is set, the filename is encoded using UTF-8
	//if bit 11 of the general purpose bit flag is not set, the filename is encoded using Encoding.GetEncoding(437);
	if ((localFile->GeneralPurposeBitFlag & (1 << 11)))
	{
		localFile->bFileName = new char[localFile->fileNameLength + 1];
		memcpy(localFile->bFileName, localFile->bFileNameHeader, localFile->fileNameLength);
		localFile->bFileName[localFile->fileNameLength] = 0;
	}
	else
	{
		CodePage437 cp437 = CodePage437();
		localFile->bFileName = cp437.Enc(localFile->bFileNameHeader, localFile->fileNameLength);
	}



	localFile->bExtraField = new char[localFile->extraFieldLength];
	_zipFs.read((char*)localFile->bExtraField, localFile->extraFieldLength); // Extra Field

	if (localFile->extraFieldLength > 0)
		ZipExtraFieldRead::ReadExtraField(localFile, false);

	localFile->DataLocation = (unsigned long)_zipFs.tellg();

	if (localFile->GeneralPurposeBitFlag & 8)
	{
		_zipFs.seekg(compressedSize, ios::cur);
		_zipFs.read((char*)&localFile->CRC, 4); // CRC
		if (localFile->CRC == 0x08074b50)
		{
			_zipFs.read((char*)&localFile->CRC, 4); // CRC
		}

		if (localFile->IsZip64)
		{
			_zipFs.read((char*)&localFile->CompressedSize, 8); // Compressed Size
			_zipFs.read((char*)&localFile->UncompressedSize, 8); // Uncompressed Size
		}
		else
		{
			_zipFs.read((char*)&tUInt, 4); // Compressed Size
			localFile->CompressedSize = tUInt;

			_zipFs.read((char*)&tUInt, 4); // Uncompressed Size
			localFile->UncompressedSize = tUInt;
		}
	}
	return ZipGood;
}

File: ZipReaderLib/Zip.cs
using System.Text;

namespace ZipReaderLib
{
    public partial class Zip
    {
        private uint _localFilesCount;
        private ZipHeader[] _centralDirectoryHeaders;
        private ZipHeader[] _localFileHeaders;
        public int LocalFilesCount => _localFileHeaders.Length;

        public ZipHeader GetCentralFile(int i)
        {
            return _centralDirectoryHeaders[i];
        }
        public ZipHeader GetLocalFile(int i)
        {
            return _localFileHeaders[i];
        }



        private byte[] _fileComment;
        public string FileComment => CompressUtils.GetString(_fileComment);



        private Stream _zipFs;
        internal ulong _centralDirStart;
        internal ulong _centralDirSize;
        private ulong _endOfCentralDir64;

        private bool _zip64;

        internal ulong offset = 0;
        internal bool ExtraDataFoundOnEndOfFile = false;


        public Zip()
        {
            CompressUtils.EncodeSetup();
        }

        #region OpenClose

        public ZipReturn ZipFileOpen(string newFilename)
        {
            ZipFileClose();
            Stream inStream;
            try
            {
                if (!File.Exists(newFilename))
                {
                    ZipFileClose();
                    return ZipReturn.ZipErrorFileNotFound;
                }

                inStream = File.OpenRead(newFilename);
            }
            catch (PathTooLongException)
            {
                ZipFileClose();
                return ZipReturn.ZipFileNameToLong;
            }
            catch (Exception)
            {
                ZipFileClose();
                return ZipReturn.ZipErrorOpeningFile;
            }

            return ZipFileOpen(inStream);
        }


        public ZipReturn ZipFileOpen(Stream inStream)
        {
            ZipFileClose();
            _zip64 = false;
            _centralDirStart = 0;
            _centralDirSize = 0;
            _zipFs = inStream;

            ZipReturn retV = ZipFileReadHeaders();
            ZipFileClose();
            return retV;
        }


        public void ZipFileClose()
        {
            if (_zipFs != null)
            {
                _zipFs.Close();
                _zipFs.Dispose();
            }
        }
        #endregion


        #region MainBodyReader

        private ZipReturn ZipFileReadHeaders()
        {
            try
            {
                ZipReturn zRet = FindEndOfCentralDirSignature();
                if (zRet != ZipReturn.ZipGood)
                {
                    return zRet;
                }

                ulong endOfCentralDir = (ulong)_zipFs.Position;
                zRet = EndOfCentralDirRead();
                if (zRet != ZipReturn.ZipGood)
                {
                    return zRet;
                }

                // check if ZIP64 header is required
                bool zip64Required = (_centralDirStart == 0xffffffff || _centralDirSize == 0xffffffff || _localFilesCount == 0xffff);

                // check for a ZIP64 header
                _zipFs.Position = (long)endOfCentralDir - 20;
                zRet = Zip64EndOfCentralDirectoryLocatorRead();
                if (zRet == ZipReturn.ZipGood)
                {
                    _zipFs.Position = (long)_endOfCentralDir64;
                    zRet = Zip64EndOfCentralDirRead();
                    if (zRet == ZipReturn.ZipGood)
                    {
                        _zip64 = true;
                        endOfCentralDir = _endOfCentralDir64;
                    }
                }

                if (zip64Required && !_zip64)
                    return ZipReturn.Zip64EndOfCentralDirError;

                offset = (endOfCentralDir - _centralDirSize) - _centralDirStart;

                _centralDirStart += offset;

                // now read the central directory
                _zipFs.Position = (long)_centralDirStart;

                _centralDirectoryHeaders = new ZipHeader[_localFilesCount];
                for (int i = 0; i < _localFilesCount; i++)
                {
                    zRet = CentralDirectoryHeaderRead(offset, out ZipHeader CentralFile);
                    if (zRet != ZipReturn.ZipGood)
                        return zRet;

                    _centralDirectoryHeaders[i] = CentralFile;
                }

                _localFileHeaders = new ZipHeader[_localFilesCount];
                for (int i = 0; i < _localFilesCount; i++)
                {
                    zRet = LocalFileHeaderRead(_centralDirectoryHeaders[i].RelativeOffsetOfLocalHeader, _centralDirectoryHeaders[i].CompressedSize, out ZipHeader LocalFile);
                    if (zRet != ZipReturn.ZipGood)
                        return zRet;

                    _localFileHeaders[i] = LocalFile;
                }

                return ZipReturn.ZipGood;
            }
            catch
            {
                return ZipReturn.ZipErrorReadingFile;
            }
        }
        #endregion


        #region EndOfCentralDirReaders
        private const uint EndOfCentralDirSignature = 0x06054b50;
        private const uint Zip64EndOfCentralDirSignature = 0x06064b50;
        private const uint Zip64EndOfCentralDirectoryLocator = 0x07064b50;

        private ZipReturn FindEndOfCentralDirSignature()
        {
            long fileSize = _zipFs.Length;
            long maxBackSearch = 0xffff;

            if (_zipFs.Length < maxBackSearch)
            {
                maxBackSearch = fileSize;
            }

            const long buffSize = 0x400;

            byte[] buffer = new byte[buffSize + 4];

            long backPosition = 4;
            while (backPosition < maxBackSearch)
            {
                backPosition += buffSize;
                if (backPosition > maxBackSearch)
                {
                    backPosition = maxBackSearch;
                }

                long readSize = backPosition > buffSize + 4 ? buffSize + 4 : backPosition;

                _zipFs.Position = fileSize - backPosition;

                _zipFs.Read(buffer, 0, (int)readSize);


                for (long i = readSize - 4; i >= 0; i--)
                {
                    if (buffer[i] != 0x50 || buffer[i + 1] != 0x4b || buffer[i + 2] != 0x05 || buffer[i + 3] != 0x06)
                    {
                        continue;
                    }

                    _zipFs.Position = fileSize - backPosition + i;
                    return ZipReturn.ZipGood;
                }
            }
            return ZipReturn.ZipCentralDirError;
        }


        private ZipReturn EndOfCentralDirRead()
        {
            using BinaryReader zipBr = new(_zipFs, Encoding.UTF8, true);
            uint thisSignature = zipBr.ReadUInt32();
            if (thisSignature != EndOfCentralDirSignature)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            ushort tUShort = zipBr.ReadUInt16(); // NumberOfThisDisk
            if (tUShort != 0)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            tUShort = zipBr.ReadUInt16(); // NumberOfThisDiskCenterDir
            if (tUShort != 0)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            _localFilesCount = zipBr.ReadUInt16(); // TotalNumberOfEntriesDisk

            tUShort = zipBr.ReadUInt16(); // TotalNumber of entries in the central directory 
            if (tUShort != _localFilesCount)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            _centralDirSize = zipBr.ReadUInt32(); // SizeOfCentralDir
            _centralDirStart = zipBr.ReadUInt32(); // Offset

            ushort zipFileCommentLength = zipBr.ReadUInt16();

            _fileComment = zipBr.ReadBytes(zipFileCommentLength);

            ExtraDataFoundOnEndOfFile = _zipFs.Position != _zipFs.Length;

            return ZipReturn.ZipGood;
        }


        private ZipReturn Zip64EndOfCentralDirRead()
        {
            using BinaryReader zipBr = new(_zipFs, Encoding.UTF8, true);
            uint thisSignature = zipBr.ReadUInt32();
            if (thisSignature != Zip64EndOfCentralDirSignature)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            ulong tULong = zipBr.ReadUInt64(); // Size of zip64 end of central directory record
            if (tULong != 44)
            {
                return ZipReturn.Zip64EndOfCentralDirError;
            }

            zipBr.ReadUInt16(); // version made by

            ushort tUShort = zipBr.ReadUInt16(); // version needed to extract
            if (tUShort != 45)
            {
                return ZipReturn.Zip64EndOfCentralDirError;
            }

            uint tUInt = zipBr.ReadUInt32(); // number of this disk
            if (tUInt != 0)
            {
                return ZipReturn.Zip64EndOfCentralDirError;
            }

            tUInt = zipBr.ReadUInt32(); // number of the disk with the start of the central directory
            if (tUInt != 0)
            {
                return ZipReturn.Zip64EndOfCentralDirError;
            }

            _localFilesCount = (uint)zipBr.ReadUInt64(); // total number of entries in the central directory on this disk

            tULong = zipBr.ReadUInt64(); // total number of entries in the central directory
            if (tULong != _localFilesCount)
            {
                return ZipReturn.Zip64EndOfCentralDirError;
            }

            _zip64 = true;
            _centralDirSize = zipBr.ReadUInt64(); // size of central directory

            _centralDirStart = zipBr.ReadUInt64(); // offset of start of central directory with respect to the starting disk number

            return ZipReturn.ZipGood;
        }


        private ZipReturn Zip64EndOfCentralDirectoryLocatorRead()
        {
            using BinaryReader zipBr = new(_zipFs, Encoding.UTF8, true);
            uint thisSignature = zipBr.ReadUInt32();
            if (thisSignature != Zip64EndOfCentralDirectoryLocator)
            {
                return ZipReturn.ZipEndOfCentralDirectoryError;
            }

            uint tUInt = zipBr.ReadUInt32(); // number of the disk with the start of the zip64 end of central directory
            if (tUInt != 0)
            {
                return ZipReturn.Zip64EndOfCentralDirectoryLocatorError;
            }

            _endOfCentralDir64 = zipBr.ReadUInt64(); // relative offset of the zip64 end of central directory record

            tUInt = zipBr.ReadUInt32(); // total number of disks
            if (tUInt > 1)
            {
                return ZipReturn.Zip64EndOfCentralDirectoryLocatorError;
            }

            return ZipReturn.ZipGood;
        }


        #endregion


        #region HeaderReaders
        private const uint LocalFileHeaderSignature = 0x04034b50;
        private const uint CentralDirectoryHeaderSignature = 0x02014b50;

        internal ZipReturn CentralDirectoryHeaderRead(ulong offset, out ZipHeader centralFile)
        {
            try
            {
                centralFile = new ZipHeader();
                using BinaryReader br = new(_zipFs, Encoding.UTF8, true);
                uint thisSignature = br.ReadUInt32();
                if (thisSignature != CentralDirectoryHeaderSignature)
                    return ZipReturn.ZipCentralDirError;

                centralFile.VersionMadeBy = br.ReadUInt16(); // Version Made By
                centralFile.VersionNeededToExtract = br.ReadUInt16(); // Version Needed To Extract

                centralFile.GeneralPurposeBitFlag = br.ReadUInt16();
                centralFile.CompressionMethod = br.ReadUInt16();

                ushort lastModFileTime = br.ReadUInt16();
                ushort lastModFileDate = br.ReadUInt16();

                centralFile.HeaderLastModified = CompressUtils.CombineDosDateTime(lastModFileDate, lastModFileTime);

                centralFile.CRC = ReadCRC(br);

                centralFile.CompressedSize = br.ReadUInt32();
                centralFile.UncompressedSize = br.ReadUInt32();

                ushort fileNameLength = br.ReadUInt16();
                ushort extraFieldLength = br.ReadUInt16();
                ushort fileCommentLength = br.ReadUInt16();

                br.ReadUInt16(); // diskNumberStart
                br.ReadUInt16(); // internalFileAttributes
                br.ReadUInt32(); // externalFileAttributes

                centralFile.RelativeOffsetOfLocalHeader = br.ReadUInt32();


                centralFile.bFileName = br.ReadBytes(fileNameLength);
                centralFile.Filename = (centralFile.GeneralPurposeBitFlag & (1 << 11)) == 0
                    ? CompressUtils.GetString(centralFile.bFileName)
                    : Encoding.UTF8.GetString(centralFile.bFileName, 0, fileNameLength);

                centralFile.IsZip64 = false;
                centralFile.ExtraDataFound = false;
                if (extraFieldLength > 0)
                {
                    centralFile.bExtraField = br.ReadBytes(extraFieldLength);
                    ZipReturn zr = ZipExtraFieldRead.ExtraFieldRead(centralFile, true);
                    if (zr != ZipReturn.ZipGood)
                        return zr;
                }

                centralFile.RelativeOffsetOfLocalHeader += offset;

                if (fileCommentLength > 0)
                {
                    centralFile.bFileComment = br.ReadBytes(fileCommentLength);
                }

                return ZipReturn.ZipGood;
            }
            catch
            {
                centralFile = null;
                return ZipReturn.ZipCentralDirError;
            }
        }

        internal ZipReturn LocalFileHeaderRead(ulong RelativeOffsetOfLocalHeader, ulong CompressedSize, out ZipHeader localFile)
        {
            try
            {
                localFile = new ZipHeader();
                using (BinaryReader br = new(_zipFs, Encoding.UTF8, true))
                {
                    localFile.RelativeOffsetOfLocalHeader = RelativeOffsetOfLocalHeader;
                    zipFs.Position = (long)RelativeOffsetOfLocalHeader;
                    uint thisSignature = br.ReadUInt32();
                    if (thisSignature != LocalFileHeaderSignature)
                        return ZipReturn.ZipLocalFileHeaderError;

                    localFile.VersionNeededToExtract = br.ReadUInt16(); // version needed to extract
                    localFile.GeneralPurposeBitFlag = br.ReadUInt16();

                    localFile.CompressionMethod = br.ReadUInt16();

                    ushort lastModFileTime = br.ReadUInt16();
                    ushort lastModFileDate = br.ReadUInt16();

                    localFile.HeaderLastModified = CompressUtils.CombineDosDateTime(lastModFileDate, lastModFileTime);

                    localFile.CRC = ReadCRC(br);
                    localFile.CompressedSize = br.ReadUInt32();
                    localFile.UncompressedSize = br.ReadUInt32();
                    ulong localRelativeOffset = 0;

                    ushort fileNameLength = br.ReadUInt16();
                    ushort extraFieldLength = br.ReadUInt16();


                    localFile.bFileName = br.ReadBytes(fileNameLength);
                    localFile.Filename = (localFile.GeneralPurposeBitFlag & (1 << 11)) == 0
                        ? CompressUtils.GetString(localFile.bFileName)
                        : Encoding.UTF8.GetString(localFile.bFileName, 0, fileNameLength);

                    localFile.IsZip64 = false;
                    localFile.ExtraDataFound = false;
                    if (extraFieldLength > 0)
                    {
                        localFile.bExtraField = br.ReadBytes(extraFieldLength);

                        ZipReturn zr = ZipExtraFieldRead.ExtraFieldRead(localFile, false);
                        if (zr != ZipReturn.ZipGood)
                            return zr;
                    }


                    localFile.DataLocation = (ulong)zipFs.Position;

                    if ((localFile.GeneralPurposeBitFlag & 8) == 8)
                    {
                        zipFs.Position += (long)CompressedSize;

                        localFile.CRC = ReadCRC(br);
                        if (CompressUtils.ByteArrCompare(localFile.CRC, new byte[] { 0x08, 0x07, 0x4b, 0x50 }))
                        {
                            localFile.CRC = ReadCRC(br);
                        }

                        if (localFile.IsZip64)
                        {
                            localFile.CompressedSize = br.ReadUInt64();
                            localFile.UncompressedSize = br.ReadUInt64();
                        }
                        else
                        {
                            localFile.CompressedSize = br.ReadUInt32();
                            localFile.UncompressedSize = br.ReadUInt32();
                        }
                    }
                    return ZipReturn.ZipGood;
                }
            }
            catch
            {
                localFile = null;
                return ZipReturn.ZipLocalFileHeaderError;
            }
        }

        private static byte[] ReadCRC(BinaryReader br)
        {
            byte[] tCRC = new byte[4];
            tCRC[3] = br.ReadByte();
            tCRC[2] = br.ReadByte();
            tCRC[1] = br.ReadByte();
            tCRC[0] = br.ReadByte();
            return tCRC;
        }


        #endregion
    }
}
